### PR TITLE
Reduce compile time

### PR DESCRIPTION
**Pull request by antmat**
Several improvements in granularity:
  * split api classes and their category_traits - traits are only used in instantiation via repository and has nothing to do with interfaces.
  * split basic_dispatch and tagged dispatch. dispatch is very heavy, so use tagged one only when required.
  * get rid of karma - so compilation boost, much faster, wow.
  * use synchronization inside registrar impl, not outside.

### CODE REVIEW

**Comment by antmat** [comment]
@3Hren review?

**Comment by 3Hren** [comment]
I see much code has been moved somewhere. Seems like you've not added new files, huh?

**Comment by antmat** [comment]
oops, fixed

File: include/cocaine/repository/cluster.hpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        ptr_type
+        get(context_t& context, interface& locator, const std::string& name, const dynamic_t& args) = 0;
+    };
+
+    template<class T>
+    struct default_factory: public factory_type {
+        virtual
+        ptr_type
+        get(context_t& context, interface& locator, const std::string& name, const dynamic_t& args) {
+        return ptr_type(new T(context, locator, name, args));
```
Comment (3Hren) [inline include/cocaine/repository/cluster.hpp]: !

File: include/cocaine/repository/gateway.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        ptr_type
+        get(context_t& context, const std::string& name, const dynamic_t& args) = 0;
+    };
+
+    template<class T>
+    struct default_factory: public factory_type {
+        virtual
+        ptr_type
+        get(context_t& context, const std::string& name, const dynamic_t& args) {
+        return ptr_type(new T(context, name, args));
```
Comment (3Hren) [inline include/cocaine/repository/gateway.hpp]: !

**Pull request merged**